// @angular-architects/module-federation-tools/esm2022/lib/mf-tools.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { WebComponentWrapper } from './web-components/web-component-wrapper';
import * as i0 from "@angular/core";
export class ModuleFederationToolsModule {
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.0.2", ngImport: i0, type: ModuleFederationToolsModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.0.2", ngImport: i0, type: ModuleFederationToolsModule, declarations: [WebComponentWrapper], imports: [CommonModule], exports: [WebComponentWrapper] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.0.2", ngImport: i0, type: ModuleFederationToolsModule, imports: [CommonModule] }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.0.2", ngImport: i0, type: ModuleFederationToolsModule, decorators: [{
            type: NgModule,
            args: [{
                    imports: [CommonModule],
                    declarations: [WebComponentWrapper],
                    exports: [WebComponentWrapper],
                }]
        }] });
